[docs] Improve documentation of the system (#23294)
mui-org/material-ui: docs/src/pages/system/sx/sx.md

# The `sx` prop

<p class="description">Utility for styling elements inline, using the theme values.</p>

## Getting Started

> Each `@material-ui/core` component supports the `sx` prop. The prop is a superset of CSS, that let's you add any CSS to the underlaying component, while mapping values to different theme keys. This should help you to keep your application look consistent.

Reviewer: I'm not sure I would start with mentioning `@material-ui/core`.

Author: What do you think of this:

> The `sx` prop is a superset of CSS, that let's you add any CSS to the underlaying component, while mapping values to different theme keys. This should help you to keep your application look consistent. The prop is available in all `@material-ui/core` components. 

Later when we support this on all primitives, we can add that here too. Or at least, we can export the `styleFunctionSx` and add something like:

> You can include this prop on your own components too by using the `styleFunctionSx` style function from `@material-ui/core/Box`, or you can automatically add it to your component by using the `experimentalStyled` from `@material-ui/core/styles`.
> ```
> import { styleFunctionSx } from '@material-ui/core/Box';
> import styled from '@emotion/styled';
> 
> const Div = styled('div')(styleFunctionSx);
> ```
> or
> ```
> import { experimentalStyled as styled } from '@material-ui/core/styles';
> 
> const Div = styled('div')``;
> ```

Reviewer 2: I'l wait for it to be updated before correcting this.

Reviewer 3: We haven't decided on adding `sx` to every component nor have we released it. We shouldn't document something before.

Reviewer: > We haven't decided on adding sx to every component nor have we released it. We shouldn't document something before.

The Slider is meant to be a projection of how all the other components will be architectured. Slider now supports the sx prop and was released. So I think that in theory, this decision and release event has already happened. I would argue that the decision started 2 years ago with

> In the future, we will look into using it directly in the core components.

https://medium.com/material-ui/introducing-material-ui-design-system-93e921beb8df

"look into" as how can we make it fast enough. Then about a year ago with (#15561).

Having a look at all the issues that where closed linking to this issue is interesting to gauge the pain.

Reviewer 3: > So I think that in theory, this decision and release event has already happened. 

That is not how an alpha works. Its goal is to give us freedom to add and remove things without having to release a major version every time. If you're now flipped to "everything released in alpha is stable" then you should've told us that before we started working on v5. That's a pretty fundamental change to the alpha process especially if it involves decisions we haven't discussed.

As ususal: something existing is in no way shape or form under any circumstance and argument for its existence. Slider with sx is not a good API just because it shipped with it. The original PR still doesn't include a real world use case and just some toying around.

> I would argue that the decision started 2 years ago with

A decision isn't a process. Could you link me the conclusion of the discussion with having widgets with system props that just affect the root container?

Reviewer: > Its goal is to give us freedom to add and remove things without having to release a major version every time.

Agree, the objective of pre-releases is to allow a short feedback loop. It allows us to validate ideas directly with developers. On this one, we are waiting to get insight from developers after having released it. It seems that the trend of improvement is toward supporting `sx` everywhere: #23220. The value of the prop increases the more it can be used in a systematic manner.

---

I don't understand the concern with the `sx` prop, the value proposition started to be discussed in #22819, then #23053, e.g. https://github.com/mui-org/material-ui/pull/23053#issuecomment-710095357.

Reviewer 3: > I don't understand the concern with the sx prop

It's JavaScript. JavaScript is only good if it solves a problem. On the flipside: I do not understand how you don't see a problem with adding features after maintaining open-source for 3+ years.

> We recommend you use this prop whenever you need to add one-of style overrides on the Material-UI components. If you repetedly add the same styles on some component, than the `styled()` is better alternative, as it allows you to specify the overrides only ones and reuse them in all component instances.

Author: I'll work next on the `styled()` utility docs page, will link it here once it's done

> We recommend you use this prop whenever you need to add one-of style overrides on the Material-UI components. If you repetedly add the same styles on some component, than the `styled()` is better alternative, as it allows you to specify the overrides only ones and reuse them in all component instances.

Reviewer 2: ```suggestion
You should use this prop whenever you need to add a style override to a Material-UI component. If you repeatedly add the same styles on a component, then `styled()` is better alternative, as it allows you to specify the overrides only once, and reuse them in all component instances.
```

{{"demo": "pages/system/sx/GettingStarted.js", "defaultCodeOpen": true}}

# Property to theme key mapping

> You can explore the [System API](system/api/) page to discover how the different css (and custom) properties are mapped to the theme keys.

Author: Should we inline here the table from `system/api`?

> You can explore the [System API](/system/api/) page to discover how the different css (and custom) properties are mapped to the theme keys.

Reviewer 2: ```suggestion
You can explore the [System API](/system/api/) page to discover how the different CSS (and custom) properties are mapped to the theme keys.
```

## Value as a function

> If you wish to use the theme for some css property that is not supported by the system, you can use function as a value, where you can have access to the theme object.

Reviewer 2: ```suggestion
If you wish to use the theme for a CSS property that is not supported by the system, you can use a function as the value, in which you can  access the theme object.
```

{{"demo": "pages/system/sx/ValueAsFunction.js", "defaultCodeOpen": true}}

# Breakpoints

> If you would like to have responsive values for some CSS property, you may use the breakpoints shorthand syntax. We offer two ways of defining the breakpoints:

Reviewer 2: ```suggestion
If you would like to have responsive values for a CSS property, you can use the breakpoints shorthand syntax. There are two ways of defining the breakpoints:
```

> ## Breakpoints as array

Reviewer 2: ```suggestion
## Breakpoints as an array
```
Should we number these?

```suggestion
## 1. Breakpoints as an array
```

The first option is to define your breakpoints as an array, from the smallest to the largest breakpoint.

{{"demo": "pages/system/sx/BreakpointsAsArray.js", "defaultCodeOpen": true}}

> ## Breakpoints as object

Reviewer 2: ```suggestion
## Breakpoints as an object
```
```suggestion
## 2. Breakpoints as an object
```

> The second option for defining breakpoints is to define them as object, using the breakpoints as keys. Here is an example of how you may define the previous example by using the object syntax.

Reviewer 2: ```suggestion
The second option for defining breakpoints is to define them as an object, using the breakpoints as keys. Here is  the previous example again, using the object syntax.
```

{{"demo": "pages/system/sx/BreakpointsAsObject.js", "defaultCodeOpen": true}}

## Custom breakpoints

> You can also specify your own custom breakpoints and use them as keys when defining the breakpoints object. Here is an example of how to do that.

Reviewer 2: ```suggestion
You can also specify your own custom breakpoints, and use them as keys when defining the breakpoints object. Here is an example of how to do that.
```
